### PR TITLE
chore: Custom Roles to manage permissions [ UI ]

### PR DESCRIPTION
**Pull request by sojan-official**
In admin settings, this Pr will add the UI for managing custom roles  ( ref: https://github.com/chatwoot/chatwoot/pull/9995 ).  It also handles the routing logic changes to accommodate fine-tuned permissions.

### CODE REVIEW

File: app/javascript/dashboard/constants/permissions.js [hunk 1]
```diff
@@ -0,0 +1,50 @@
+export const AVAILABLE_CUSTOM_ROLE_PERMISSIONS = [
+  'conversation_manage',
+  'conversation_unassigned_manage',
+  'conversation_participating_manage',
+  'contact_manage',
+  'report_manage',
+  'knowledge_base_manage',
+];
+
+export const ROLE_PERMISSIONS = ['agent', 'administrator'];
```
Comment (muhsin-k) [inline app/javascript/dashboard/constants/permissions.js:10]: Lets use `ROLES`
```suggestion
export const ROLES = ['agent', 'administrator'];
```

File: app/javascript/dashboard/routes/dashboard/settings/agents/AddAgent.vue [hunk 2]
```diff
@@ -1,156 +1,164 @@
...
+          error: errorResponse = '',
+          attributes: attributes = [],
+          message: attrError = '',
+        } = {},
+      } = {},
+    } = error;
+
+    let errorMessage = '';
+    if (error?.response?.status === 422 && !attributes.includes('base')) {
+      errorMessage = this.$t('AGENT_MGMT.ADD.API.EXIST_MESSAGE');
```
Comment (muhsin-k) [inline app/javascript/dashboard/routes/dashboard/settings/agents/AddAgent.vue:96]: `this` will not work here.

File: app/javascript/dashboard/routes/dashboard/settings/agents/AddAgent.vue [hunk 3]
```diff
@@ -1,156 +1,164 @@
...
+    } else {
+      errorMessage = this.$t('AGENT_MGMT.ADD.API.ERROR_MESSAGE');
```
Comment (muhsin-k) [inline app/javascript/dashboard/routes/dashboard/settings/agents/AddAgent.vue:98]: `this` will not work here.

File: app/javascript/dashboard/routes/dashboard/settings/agents/Index.vue [hunk 4]
```diff
@@ -34,11 +38,23 @@ const deleteMessage = computed(() => {
...
 const currentUserId = computed(() => getters.getCurrentUserID.value);
+const customRoles = useMapGetter('customRole/getCustomRoles');
 
 onMounted(() => {
   store.dispatch('agents/get');
+  store.dispatch('customRole/getCustomRole');
 });
 
+const getAgentRoleName = agent => {
+  if (agent.custom_role_id === null) {
```
Comment (muhsin-k) [inline app/javascript/dashboard/routes/dashboard/settings/agents/Index.vue:49]:
```suggestion
  if (!agent.custom_role_id) {
```

File: app/javascript/dashboard/routes/dashboard/settings/customRoles/Index.vue [hunk 5]
```diff
@@ -0,0 +1,193 @@
...
+    // Ignore Error
+  }
+};
+
+onMounted(() => {
+  fetchCustomRoles();
+});
+
+const showAlertMessage = message => {
+  loading[activeResponse.value.id] = false;
```
Comment (muhsin-k) [inline app/javascript/dashboard/routes/dashboard/settings/customRoles/Index.vue:64]: Should we check loading.value here. Please check this condition.
```suggestion
loading.value[activeResponse.value.id] = false;
```

File: app/javascript/dashboard/routes/dashboard/settings/customRoles/Index.vue [hunk 6]
```diff
@@ -0,0 +1,193 @@
...
+  activeResponse.value = {};
+  customRoleAPI.value.message = message;
```
Comment (muhsin-k) [inline app/javascript/dashboard/routes/dashboard/settings/customRoles/Index.vue:66]: Unused reference variable.

File: app/javascript/dashboard/routes/dashboard/settings/customRoles/Index.vue [hunk 7]
```diff
@@ -0,0 +1,193 @@
...
+  useAlert(message);
+};
+
+const openAddModal = () => {
+  if (isBehindAPaywall.value) {
+    return;
+  }
```
Comment (muhsin-k) [inline app/javascript/dashboard/routes/dashboard/settings/customRoles/Index.vue:73]:
```suggestion
if (isBehindAPaywall.value) return;
```

File: app/javascript/dashboard/routes/dashboard/settings/customRoles/component/CustomRolePaywall.vue [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+
+const isOnChatwootCloud = useMapGetter('globalConfig/isOnChatwootCloud');
+
+const currentUser = useMapGetter('getCurrentUser');
+const currentAccountId = useMapGetter('getCurrentAccountId');
+
+const isSuperAdmin = computed(() => {
+  return currentUser.value.type === 'SuperAdmin';
+});
+const i18nKey = isOnChatwootCloud.value ? 'PAYWALL' : 'ENTERPRISE_PAYWALL';
```
Comment (muhsin-k) [inline app/javascript/dashboard/routes/dashboard/settings/customRoles/component/CustomRolePaywall.vue:52]: Should we use a computed property for better reactivity?
```suggestion
const i18nKey = computed(() => isOnChatwootCloud.value ? 'PAYWALL' : 'ENTERPRISE_PAYWALL');

```

**Comment by iamsivin** [comment]
**TODO**
1. Adding an agent with a custom role is not working (API).
2. When editing an agent's role from a custom role to agent or admin, the custom role is not being reset to nill (API).
3. Help center article edit/create actions are not functioning properly and show the error: `You are not authorized to do this action` (API).

CC @sojan-official

**Comment by muhsin-k** [comment]
Let's display the profile settings for all users, allowing them to update basic profile details such as name, profile picture, and password.
CleanShot 2024-09-11 at 00 06 23
cc @sojan-official

**Comment by sojan-official** [comment]
> **TODO**
> 
> 1. Adding an agent with a custom role is not working (API).
> 2. When editing an agent's role from a custom role to agent or admin, the custom role is not being reset to nill (API).
> 3. Help center article edit/create actions are not functioning properly and show the error: `You are not authorized to do this action` (API).
> 
> CC @sojan-official

This is fixed.

**Comment by iamsivin** [comment]
> @iamsivin Let's add the ability to open the profile settings for all the roles, allowing them to update basic profile details such as name, profile picture, and password. 

@muhsin-k This is fixed

**Review by muhsin-k — APPROVED** [review]

**Pull request merged**